File: CloudFormation/src/main/java/com/intellij/aws/cloudformation/inspections/FormatViolationInspection.java
package com.intellij.aws.cloudformation.inspections;

import com.intellij.aws.cloudformation.CloudFormationFormatChecker;
import com.intellij.aws.cloudformation.CloudFormationPsiUtils;
import com.intellij.codeInspection.InspectionManager;
import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.psi.PsiFile;
import com.intellij.util.ArrayUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class FormatViolationInspection extends LocalInspectionTool {
  @Nullable
  @Override
  public ProblemDescriptor[] checkFile(@NotNull PsiFile file, @NotNull InspectionManager manager, boolean isOnTheFly) {
    if (!CloudFormationPsiUtils.isCloudFormationFile(file)) {
      return null;
    }

    final CloudFormationFormatChecker checker = new CloudFormationFormatChecker(manager, isOnTheFly);
    checker.file(file);
    return ArrayUtil.toObjectArray(checker.getProblems(), ProblemDescriptor.class);
  }
}

File: CloudFormation/src/main/kotlin/com/intellij/aws/cloudformation/inspections/FormatViolationInspection.kt
package com.intellij.aws.cloudformation.inspections

import com.intellij.aws.cloudformation.CloudFormationInspections
import com.intellij.aws.cloudformation.CloudFormationParser
import com.intellij.aws.cloudformation.CloudFormationPsiUtils
import com.intellij.codeInspection.InspectionManager
import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.psi.PsiFile


abstract class FormatViolationInspection : LocalInspectionTool() {
  override fun runForWholeFile(): Boolean = true

  override fun checkFile(file: PsiFile, manager: InspectionManager, isOnTheFly: Boolean): Array<ProblemDescriptor>? {
    if (!CloudFormationPsiUtils.isCloudFormationFile(file)) {
      return null
    }

    val parsed = CloudFormationParser.parse(file)
    val inspected = CloudFormationInspections.inspectFile(parsed)

    val problems = parsed.problems.plus(inspected.problems).map {
      manager.createProblemDescriptor(
          it.element,
          it.description,
          isOnTheFly,
          LocalQuickFix.EMPTY_ARRAY,
          ProblemHighlightType.GENERIC_ERROR_OR_WARNING)
    }

    return problems.toTypedArray()
  }
}

class JsonFormatViolationInspection: FormatViolationInspection()
class YamlFormatViolationInspection: FormatViolationInspection()
